File: yologram-api-v2/app/domain/ums/user_email_verification_service.py
import random
from datetime import datetime, timedelta

from sqlalchemy.orm import Session

from app.core.exception import (
    UserEmailVerificationExpiredException,
    UserEmailVerificationInvalidException,
    UserDuplicateException,
)
from app.domain.ums.email_sender import EmailSender
from app.domain.ums.model import UserEmailVerification
from app.domain.ums.repository import UserEmailVerificationRepository, UserRepository


class UserEmailVerificationService:

    def __init__(self, db: Session, email_sender: EmailSender):
        self.repository = UserEmailVerificationRepository(db)
        self.user_repository = UserRepository(db)
        self.email_sender = email_sender

    def send_code(self, email: str) -> None:
        if self.user_repository.find_by_email(email):
            raise UserDuplicateException()

        self.repository.delete_by_email(email)

        code = f"{random.randint(0, 999999):06d}"
        entity = UserEmailVerification(
            email=email,
            code=code,
            expired_at=datetime.now() + timedelta(minutes=5),
        )
        self.repository.save(entity)
        self.email_sender.send_verification_code(email, code)

    def verify_code(self, email: str, code: str) -> None:
        entity = self.repository.find_latest_by_email(email)
        if not entity:
            raise UserEmailVerificationInvalidException()
        if entity.expired_at < datetime.now():
            raise UserEmailVerificationExpiredException()
        if entity.code != code:
            raise UserEmailVerificationInvalidException()
        entity.verified = True

File: yologram-api-v1/src/main/kotlin/link/yologram/api/v1/domain/ums/service/UserEmailVerificationService.kt
package link.yologram.api.v1.domain.ums.service

import link.yologram.api.v1.domain.ums.entity.UserEmailVerification
import link.yologram.api.v1.domain.ums.exception.UserEmailVerificationExpiredException
import link.yologram.api.v1.domain.ums.exception.UserEmailVerificationInvalidException
import link.yologram.api.v1.domain.ums.exception.UserDuplicateException
import link.yologram.api.v1.domain.ums.repository.UserEmailVerificationRepository
import link.yologram.api.v1.domain.ums.repository.UserRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.LocalDateTime
import kotlin.random.Random

@Service
class UserEmailVerificationService(
    private val emailVerificationCodeRepository: UserEmailVerificationRepository,
    private val userRepository: UserRepository,
    private val emailSender: EmailSender,
) {

    @Transactional
    fun sendCode(email: String) {
        if (userRepository.existsByEmail(email)) {
            throw UserDuplicateException()
        }

        emailVerificationCodeRepository.deleteAllByEmail(email)

        val code = generateCode()
        val verification = UserEmailVerification(
            email = email,
            code = code,
            expiredAt = LocalDateTime.now().plusMinutes(5),
        )
        emailVerificationCodeRepository.save(verification)

        emailSender.sendVerificationCode(email, code)
    }

    @Transactional
    fun verifyCode(email: String, code: String) {
        val verification = emailVerificationCodeRepository.findTopByEmailOrderByCreatedAtDesc(email)
            .orElseThrow { UserEmailVerificationInvalidException() }

        if (verification.expiredAt.isBefore(LocalDateTime.now())) {
            throw UserEmailVerificationExpiredException()
        }

        if (verification.code != code) {
            throw UserEmailVerificationInvalidException()
        }

        verification.verified = true
    }

    private fun generateCode(): String {
        return String.format("%06d", Random.nextInt(1_000_000))
    }
}
